master/internal/resourcemanagers/agent/agent.go

```diff
@@ -496,59 +545,63 @@ func (a *agent) summarize(ctx *actor.Context) model.AgentSummary {
 		result.Label = a.agentState.Label
 		result.Enabled = a.agentState.enabled
 		result.Draining = a.agentState.draining
-		result.NumContainers = len(a.agentState.containers)
+		result.NumContainers = len(a.agentState.containerAllocation)
 	}
 
 	return result
 }
 
 func (a *agent) gatherContainersToReattach(ctx *actor.Context) []aproto.ContainerReattach {
-	result := make([]aproto.ContainerReattach, 0, len(a.agentState.containers))
-	for containerID, allocation := range a.agentState.containers {
-		resp := ctx.Ask(allocation, sproto.GetResourcesContainerState{
-			ResourcesID: sproto.ResourcesID(containerID),
-		})
-		switch {
-		case resp.Error() != nil:
-			ctx.Log().Warnf(
-				"allocation GetTaskContainerState id: %s, got error: %s", containerID, resp.Error())
-		case resp.Get() == nil:
-			ctx.Log().Warnf("allocation GetTaskContainerState id: %s, is nil", containerID)
-		default:
-			containerState := resp.Get().(cproto.Container)
-			result = append(result, aproto.ContainerReattach{Container: containerState})
-		}
+	err := a.agentState.restoreContainersField()
+	if err != nil {
+		ctx.Log().WithError(err).Warn("failed restoreContainersField in gatherContainersToReattach")
 	}
+	result := make([]aproto.ContainerReattach, 0, len(a.agentState.containerAllocation))
+
+	for _, container := range a.agentState.containerState {
+		result = append(result, aproto.ContainerReattach{Container: *container})
+	}
+
+	fmt.Println("gatherContainersToReattach: ", result)
 	return result
 }
 
-func (a *agent) handleContainersReattached(ctx *actor.Context, agentStarted *aproto.AgentStarted) {
-	ctx.Log().Debugf("agent ContainersRestored ip: %v , containers: %v",
-		a.address, agentStarted.ContainersReattached)
+func (a *agent) handleContainersReattached(
+	ctx *actor.Context, agentStarted *aproto.AgentStarted) error {
+	ctx.Log().Debugf("agent ContainersRestored ip: %v , reattached: %v, allocations: %v",
+		a.address, agentStarted.ContainersReattached, maps.Keys(a.agentState.containerState))
 
 	recovered := map[cproto.ID]aproto.ContainerReattachAck{}
 
 	for _, containerRestored := range agentStarted.ContainersReattached {
 		if containerRestored.Failure == nil {
-			_, ok := a.agentState.containers[containerRestored.Container.ID]
+			cid := containerRestored.Container.ID
+			_, ok := a.agentState.containerAllocation[cid]
 
 			if ok {
```

Comment:
style nit here an a bit elsewhere, prefer `if !ok { continue } ...`, then no indent. the whole go code review comments 'indent error(and, for me, any exception) flow' section really helps me read code quickly. feel free to tell me to go away, ha.